### ISSUE TITLE
Be able to save the state of a visualization tool?

### ISSUE THREAD

**Issue by scottx611x** [parklab/refinery-platform#1377]
I believe there was talk about being able to save the state of a visualization tool to be able to share one's findings???

Assuming that this means there should be some model that would be representative of visualization tool?

@ngehlenborg: I had this in my notes. if you could add your thoughts here that would be great. Not a high priority!

**Comment by jkmarx** [comment]
@scottx611x This question seems tightly wrapped with the tools_api. Are there additional things to discuss here?

**Comment by scottx611x** [comment]
@jkmarx definitely! We should remember this issue for our upcoming discussions about Viz tool states for sure.

Adding @mccalluc to the convo as well

**Comment by mccalluc** [comment]
I think the API that Nils has in mind is something like:
- **POST**: Given a blob (of JSON, or is that too restrictive?) the server returns a key
- **GET**: Given a key, return the blob.

and then I think typically the visualization would use the HTML5 History API to update the url.

I think we should have some visualizations that need state preserved we invest time on this: Having real users will influence design. The first step is for visualizations to capture their state, and once that's done they can represent that state in their url. When/if a key/value API becomes available, they can start using keys instead.

Some issues to consider:
- Is the API available to all users, or only those logged in?
- Are there any limits on blob size, or POST frequency, that we want to enforce?
- Does the origin of the request matter? I think in some cases visualizations might be on a separate host... but if we let any server anywhere POST, there's the potential for abuse.
- Do we store or require any metadata? Like, if we kill a particular visualization, can be wipe all associated data? Or if one user or IP is abusive, can we clear just those records?

**Comment by drj11** [comment]
I have a question:

As the author of a visualisation tool, what persistence API would I expect to use? How does that match with what refinery will provide?

My answer would be that I, as a tool author, would expect to make my own arrangements for making visualisations persist.

I think depends a lot on whether a visualisation tool is able to stand on its own or whether it is "a refinery visualisation tool".

**Comment by drj11** [comment]
Whatever URL/POST mechanism the tool is using to store blobs in refinery, refinery should be able to pass a unique URL to the tool that is signed by refinery so that refinery can validate any incoming POST request to check it is coming from a tool it has previously launched.

I suppose in that regard, it's a bit like the approve/decline app requests that you get on Twitter/Slack.

**Comment by mccalluc** [comment]
@drj11 : To the extent possible, I want to avoid run-time dependencies on Refinery for visualization containers. The key dependency is at startup where the input needs to be in a mutually agreed location. As you said, a tool author

> would expect to make my own arrangements for making visualisations persist

One big todo is supporting docker volumes, so that containers have as much disk space as they need. The details of this should be internal to django_docker_engine.

**Comment by jkmarx** [comment]
@ngehlenborg Thoughts?
